File: recipes/claude/05-sse-streaming-claude/python/stream_messages.py
"""
Server-Sent Events (SSE) Streaming with Anthropic Claude in Python.
Demonstrates both raw event iteration and the high-level client.messages.stream() helper.
Includes event type handling (text deltas, JSON deltas, usage) and robust connection handling.
"""

import os
import sys
import time
from typing import Any, Dict
from dotenv import load_dotenv
import anthropic
from anthropic import Anthropic, APIConnectionError, APIStatusError, RateLimitError

load_dotenv()

API_KEY = os.environ.get("ANTHROPIC_API_KEY")
MODEL = os.environ.get("ANTHROPIC_MODEL", "claude-3-7-sonnet-20250219")

if not API_KEY:
    print("Error: ANTHROPIC_API_KEY is not set in environment or .env file.", file=sys.stderr)
    sys.exit(1)


def demonstrate_raw_stream(client: Anthropic) -> None:
    """
    Demonstrates manual parsing of raw Server-Sent Events via client.messages.create(stream=True).
    Captures message_start, content_block_start, content_block_delta, content_block_stop,
    message_delta, and message_stop.
    """
    print("\n--- 1. Raw Stream Event Processor (Low-Level) ---")
    start_time = time.perf_counter()
    ttft = None
    accumulated_text = []

    try:
        with client.messages.create(
            model=MODEL,
            max_tokens=256,
            stream=True,
            messages=[
                {
                    "role": "user",
                    "content": "Explain SSE streaming in one concise paragraph under 50 words.",
                }
            ],
        ) as stream:
            for event in stream:
                event_type = event.type

                if event_type == "message_start":
                    msg = event.message
                    print(f"[{event_type}] ID: {msg.id}, Model: {msg.model}, Role: {msg.role}")
                    print(f"  Input tokens: {msg.usage.input_tokens}")

                elif event_type == "content_block_start":
                    print(f"[{event_type}] Index: {event.index}, Block Type: {event.content_block.type}")

                elif event_type == "content_block_delta":
                    if ttft is None:
                        ttft = (time.perf_counter() - start_time) * 1000
                        print(f"  --> Time to First Token (TTFT): {ttft:.1f}ms")
                    
                    delta = event.delta
                    if delta.type == "text_delta":
                        sys.stdout.write(delta.text)
                        sys.stdout.flush()
                        accumulated_text.append(delta.text)
                    elif delta.type == "input_json_delta":
                        print(f"  [Tool JSON Delta]: {delta.partial_json}")

                elif event_type == "content_block_stop":
                    print(f"\n[{event_type}] Block index {event.index} completed.")

                elif event_type == "message_delta":
                    delta_info = event.delta
                    usage_info = event.usage
                    print(f"[{event_type}] Stop reason: {delta_info.stop_reason}")
                    print(f"  Output tokens: {usage_info.output_tokens}")

                elif event_type == "message_stop":
                    total_ms = (time.perf_counter() - start_time) * 1000
                    print(f"[{event_type}] Stream closed cleanly in {total_ms:.1f}ms.")

                elif event_type == "ping":
                    print(f"[{event_type}] Keep-alive heartbeat received.")

    except RateLimitError as e:
        print(f"Rate limit exceeded: {e}", file=sys.stderr)
        raise
    except APIConnectionError as e:
        print(f"Network connection failed: {e}", file=sys.stderr)
        raise
    except APIStatusError as e:
        print(f"Anthropic API returned error status {e.status_code}: {e.message}", file=sys.stderr)
        raise


def demonstrate_stream_helper(client: Anthropic) -> None:
    """
    Demonstrates the recommended client.messages.stream() context manager.
    Provides ergonomic text iteration and access to the final assembled Message object.
    """
    print("\n--- 2. High-Level Stream Helper (client.messages.stream) ---")
    start_time = time.perf_counter()
    ttft = None

    try:
        with client.messages.stream(
            model=MODEL,
            max_tokens=256,
            messages=[
                {
                    "role": "user",
                    "content": "List three operational advantages of streaming over buffered JSON responses.",
                }
            ],
        ) as stream:
            for text_chunk in stream.text_stream:
                if ttft is None:
                    ttft = (time.perf_counter() - start_time) * 1000
                    print(f"[TTFT: {ttft:.1f}ms] Output stream:")
                sys.stdout.write(text_chunk)
                sys.stdout.flush()

            # Retrieve final assembled message object with accurate token usage
            final_message = stream.get_final_message()
            print("\n")
            print("--- Final Assembled Message Metadata ---")
            print(f"Message ID:   {final_message.id}")
            print(f"Stop Reason:  {final_message.stop_reason}")
            print(f"Input Tokens: {final_message.usage.input_tokens}")
            print(f"Output Tokens:{final_message.usage.output_tokens}")

    except Exception as e:
        print(f"Error during stream helper execution: {e}", file=sys.stderr)
        raise


def main() -> None:
    print(f"Connecting to Anthropic API with model: {MODEL}")
    client = Anthropic(api_key=API_KEY)

    # 1. Execute raw event stream
    demonstrate_raw_stream(client)

    # 2. Execute high-level stream helper
    demonstrate_stream_helper(client)


if __name__ == "__main__":
    main()

File: recipes/claude/05-sse-streaming-claude/typescript/stream_messages.ts
/**
 * Server-Sent Events (SSE) Streaming with Anthropic Claude in TypeScript.
 * Demonstrates both manual event listening and the high-level createMessageStream helper.
 * Covers event types, time-to-first-token instrumentation, and final message extraction.
 */

import Anthropic from "@anthropic-ai/sdk";
import * as dotenv from "dotenv";

dotenv.config();

const apiKey = process.env.ANTHROPIC_API_KEY;
const model = process.env.ANTHROPIC_MODEL || "claude-3-7-sonnet-20250219";

if (!apiKey) {
  console.error("Error: ANTHROPIC_API_KEY is not defined in environment variables or .env");
  process.exit(1);
}

const client = new Anthropic({ apiKey });

/**
 * Demonstrates low-level event iteration using async iterable over stream response.
 */
async function demonstrateRawStream(): Promise<void> {
  console.log("\n--- 1. Low-Level Event Stream (Async Iterable) ---");
  const startTime = Date.now();
  let firstTokenLogged = false;

  try {
    const stream = await client.messages.create({
      model,
      max_tokens: 256,
      stream: true,
      messages: [
        {
          role: "user",
          content: "Explain Server-Sent Events in 40 words or less.",
        },
      ],
    });

    for await (const event of stream) {
      switch (event.type) {
        case "message_start":
          console.log(`[${event.type}] Message ID: ${event.message.id}, Role: ${event.message.role}`);
          console.log(`  Input tokens: ${event.message.usage.input_tokens}`);
          break;

        case "content_block_start":
          console.log(`[${event.type}] Block index: ${event.index}, Content type: ${event.content_block.type}`);
          break;

        case "content_block_delta":
          if (!firstTokenLogged) {
            const ttft = Date.now() - startTime;
            console.log(`  --> Time to First Token (TTFT): ${ttft}ms`);
            firstTokenLogged = true;
          }

          if (event.delta.type === "text_delta") {
            process.stdout.write(event.delta.text);
          } else if (event.delta.type === "input_json_delta") {
            console.log(`  [Partial JSON]: ${event.delta.partial_json}`);
          }
          break;

        case "content_block_stop":
          console.log(`\n[${event.type}] Finished content block index: ${event.index}`);
          break;

        case "message_delta":
          console.log(`[${event.type}] Stop reason: ${event.delta.stop_reason}`);
          console.log(`  Output tokens: ${event.usage.output_tokens}`);
          break;

        case "message_stop": {
          const totalDuration = Date.now() - startTime;
          console.log(`[${event.type}] Stream closed cleanly in ${totalDuration}ms.`);
          break;
        }

        case "ping":
          console.log(`[${event.type}] Ping frame received.`);
          break;

        default:
          break;
      }
    }
  } catch (err: unknown) {
    console.error("Error during raw stream processing:", err);
    throw err;
  }
}

/**
 * Demonstrates high-level createMessageStream helper with event emitter pattern.
 */
async function demonstrateStreamHelper(): Promise<void> {
  console.log("\n--- 2. High-Level Stream Helper (createMessageStream) ---");
  const startTime = Date.now();
  let firstTokenLogged = false;

  try {
    const stream = client.messages.stream({
      model,
      max_tokens: 256,
      messages: [
        {
          role: "user",
          content: "State three key production considerations when deploying SSE in production.",
        },
      ],
    });

    // Event listener for incoming text tokens
    stream.on("text", (deltaText: string) => {
      if (!firstTokenLogged) {
        const ttft = Date.now() - startTime;
        console.log(`[TTFT: ${ttft}ms] Text Stream:`);
        firstTokenLogged = true;
      }
      process.stdout.write(deltaText);
    });

    // Await stream completion and retrieve the fully accumulated message
    const finalMessage = await stream.finalMessage();
    console.log("\n");
    console.log("--- Final Message Summary ---");
    console.log(`Message ID:    ${finalMessage.id}`);
    console.log(`Stop Reason:   ${finalMessage.stop_reason}`);
    console.log(`Input Tokens:  ${finalMessage.usage.input_tokens}`);
    console.log(`Output Tokens: ${finalMessage.usage.output_tokens}`);
  } catch (err: unknown) {
    console.error("Error in stream helper execution:", err);
    throw err;
  }
}

async function main(): Promise<void> {
  console.log(`Starting Anthropic SSE Stream test with model: ${model}`);
  await demonstrateRawStream();
  await demonstrateStreamHelper();
}

main().catch((err) => {
  console.error("Fatal execution error:", err);
  process.exit(1);
});
